File: June-LeetCoding-Challenge/25-Find-the-Duplicate-Number/Find-the-Duplicate-Number.cs
// Runtime: 100 ms, faster than 61.31% of C# online submissions for Find the Duplicate Number.
// Memory Usage: 25.2 MB, less than 88.84% of C# online submissions for Find the Duplicate Number.
public class Solution {
    public int FindDuplicate(int[] nums) {
        
        int slow = 0;
        int fast = 0;
        int finder = 0;

        while (true)
        {
            slow = nums[slow];
            fast = nums[nums[fast]];

            if (slow == fast)
                break;
        }
        
        while (true)
        {
            slow = nums[slow];
            finder = nums[finder];
            if (slow == finder)
                break;
        }
        
        return finder;
        
    }
}

File: June-LeetCoding-Challenge/25-Find-the-Duplicate-Number/Find-the-Duplicate-Number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int a = 0, b = 0;
        do{
            a = nums[a];
            b = nums[nums[b]];
        } while (a != b);
        
        b = 0;
        while (a != b){
            a = nums[a];
            b = nums[b];
        }
        return a;
    }
};